sql-to-dbsp-compiler/lib/sqllib/src/array.rs

```diff
@@ -86,3 +87,148 @@ pub fn array_append<T>(mut vector: Vec<T>, value: T) -> Vec<T> {
 pub fn array_appendN<T>(vector: Option<Vec<T>>, value: T) -> Option<Vec<T>> {
     Some(array_append(vector?, value))
 }
+
+pub fn array_compact<T>(vector: Vec<Option<T>>) -> Vec<T> {
+    vector.into_iter().flatten().collect()
+}
+
+pub fn array_compactN<T>(vector: Option<Vec<Option<T>>>) -> Option<Vec<T>> {
+    Some(array_compact(vector?))
+}
+
+pub fn array_prepend<T>(mut vector: Vec<T>, value: T) -> Vec<T> {
+    vector.insert(0, value);
+    vector
+}
+
+pub fn array_prependN<T>(vector: Option<Vec<T>>, value: T) -> Option<Vec<T>> {
+    Some(array_prepend(vector?, value))
+}
+
+pub fn sort_array<T>(mut vector: Vec<T>, ascending: bool) -> Vec<T>
+where
+    T: Ord,
+{
+    if ascending {
+        vector.sort()
+    } else {
+        vector.sort_by(|a, b| b.cmp(a))
+    };
+
+    vector
+}
+
+pub fn array_max<T>(vector: Vec<T>) -> Option<T>
+where
+    T: Ord,
+{
+    vector.into_iter().max()
+}
+
+pub fn array_min<T>(vector: Vec<T>) -> Option<T>
+where
+    T: Ord,
+{
+    vector.into_iter().min()
+}
+
+pub fn array_maxN<T>(vector: Vec<Option<T>>) -> Option<T>
+where
+    T: Ord,
+{
+    vector.into_iter().max().flatten()
+}
+
+pub fn array_minN<T>(vector: Vec<Option<T>>) -> Option<T>
+where
+    T: Ord,
+{
+    vector.into_iter().filter(|i| i.is_some()).min().flatten()
+}
+
+pub fn array_position<T>(vector: Vec<T>, element: T) -> i64
+where
+    T: Eq,
+{
+    vector
+        .into_iter()
+        .position(|x| x == element)
+        .map(|v| v + 1)
+        .unwrap_or(0) as i64
+}
+
+pub fn array_positionN<T>(vector: Vec<T>, element: Option<T>) -> Option<i64>
+where
+    T: Eq,
+{
+    Some(array_position(vector, element?))
+}
+
+pub fn array_contains<T>(vector: Vec<T>, element: T) -> bool
+where
+    T: Eq,
+{
+    vector.contains(&element)
+}
+
+pub fn array_containsN<T>(vector: Vec<T>, element: Option<T>) -> Option<bool>
```

Comment:
where is the function where vector: Option<Vec<T>>?

Reply:
I missed that previously but now am creating a macro that does this.